Sources/Client/Client/Client+Setup.swift

```diff
@@ -0,0 +1,184 @@
+//
+//  Client+Setup.swift
+//  StreamChatCore
+//
+//  Created by Alexey Bukhtin on 09/07/2019.
+//  Copyright © 2019 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+// MARK: Setup
+
+extension Client {
+    
+    /// Setup the current user with a given token.
+    ///
+    /// - Parameters:
+    ///     - user: the current user (see `User`).
+    ///     - token: a Stream Chat API token.
+    public func set(user: User, token: Token) {
+        disconnect()
+        
+        if apiKey.isEmpty || token.isEmpty {
+            logger?.log("❌ API key or token is empty: \(apiKey), \(token)", level: .error)
+            return
+        }
+        
+        self.user = user
+        setup(token: token)
+    }
+    
+    /// Setup the current user with a token provider (see `TokenProvider`).
+    ///
+    /// A token provider is a function in which you send a request to your own backend to get a Stream Chat API token.
+    /// Then you send it to the client to complete the setup with a callback function from the token provider.
+    ///
+    /// Example:
+    /// ```
+    /// Client.shared.set(user: user) { callback in
+    ///    if let url = URL(string: "https://my.backend.io/token?user_id=\(user.id)") {
+    ///        let task = URLSession.shared.dataTask(with: url) { data, response, error in
+    ///            if error == nil,
+    ///                let json = try? JSONSerialization.jsonObject(with: data),
+    ///                let token = json["token"] as? String {
+    ///                callback(token)
+    ///            } else {
+    ///                // Handle the error.
+    ///                print("Token request failed", error)
+    ///            }
+    ///        }
+    ///
+    ///        task.resume()
+    ///    }
+    /// }
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - user: the current user (see `User`).
+    ///   - tokenProvider: a token provider.
+    public func set(user: User, _ tokenProvider: @escaping TokenProvider) {
+        disconnect()
+        
+        if apiKey.isEmpty {
+            logger?.log("❌ API key is empty.", level: .error)
+            return
+        }
+        
+        self.user = user
+        self.tokenProvider = tokenProvider
+        touchTokenProvider()
```

Comment:
this is weird; you call a fn on set user that disconnects the websocket.

Reply:
We need to update Token and reconnect.